core/src/main/java/org/infinispan/interceptors/distribution/BaseDistributionInterceptor.java

```diff
@@ -378,55 +381,157 @@ public Object visitGetAllCommand(InvocationContext ctx, GetAllCommand command) t
+   }
+
+   protected CompletableFuture<Void> handleRemotelyRetrievedKeys(InvocationContext ctx, CompletableFuture<Void> future, Function<CompletableFuture<?>, List<?>> keysProvider) {
+      return future;
+   }
+
+   /**
+    * {@link #completeExceptionally(Throwable)} must be called from synchronized block since we must not complete
+    * the future (exceptionally) when we're accessing the context - if there was an exception and we would retry,
+    * the context could be accessed concurrently by dangling handlers and retry execution (that does not synchronize
+    * on the same future).
+    */
+   private class ClusteredGetAllFuture extends CompletableFuture<Void> implements InvocationSuccessFunction, Function<CompletableFuture<?>, List<?>> {
+      private GetAllCommand localCommand;
+      private int counter;
+
+      public ClusteredGetAllFuture(int counter, GetAllCommand localCommand) {
+         this.counter = counter;
+         this.localCommand = localCommand;
+      }
+
+      @Override
+      public synchronized boolean completeExceptionally(Throwable ex) {
+         return super.completeExceptionally(ex);
+      }
+
+      @Override
+      public Object apply(InvocationContext rCtx, VisitableCommand rCommand, Object rv) throws Throwable {
+         assert rv == null; // value with which the allFuture has been completed
+         return invokeNext(rCtx, localCommand);
+      }
+
+      @Override
+      public List<?> apply(CompletableFuture<?> cf) {
+         // we could use Supplier<List<?>> but that would result in a clash of CompletableFuture.get() and Supplier.get()
+         assert cf == this;
+         return new ArrayList<>(localCommand.getKeys());
+      }
+   }
+
+   private class ClusteredGetAllHandler implements BiConsumer<Map<Address, Response>, Throwable> {
+      private final ClusteredGetAllFuture allFuture;
+      private final InvocationContext ctx;
+      private final GetAllCommand command;
+      private final List<?> keys;
+      private final Collection<Address> contactedNodes;
+
+      private ClusteredGetAllHandler(ClusteredGetAllFuture allFuture, InvocationContext ctx, GetAllCommand command, List<?> keys, Collection<Address> contactedNodes) {
+         this.allFuture = allFuture;
+         this.keys = keys;
+         this.ctx = ctx;
+         this.command = command;
+         this.contactedNodes = contactedNodes;
+      }
+
+      @Override
+      public void accept(Map<Address, Response> responseMap, Throwable throwable) {
+         if (throwable != null) {
+            allFuture.completeExceptionally(throwable);
+            return;
+         }
+         Response response = getSingleReadResponseOrFail(responseMap, allFuture);
+         if (response == null) {
+            if (!allFuture.isDone()) {
+               // When we did not get any successful response but the future was not exceptionally completed
+               // that means that the node got suspected and we have to contact other owners
```

Comment:
It's not that easy as using staggered gets - you don't want to send `ClusteredGetAllCommand(keys.filter(k -> isOwner(k) == address))` to each node, because that way you would get the values multiple times. If sending a subset of keys to one node fails, you have to contact multiple other nodes...